crates/wast/src/lexer.rs

```diff
@@ -46,13 +46,14 @@ pub struct Lexer<'a> {
 
 /// A single token parsed from a `Lexer`.
 #[derive(Copy, Clone, Debug, PartialEq)]
+#[repr(C)] // ensure the assertion below always works
```

Comment:
This attribute doesn't affect the inner structs. Since those would get randomized to they can also affect layout.

You could do a bunch of runs with -Zrandomize-layout and different `-Zlayout-seed=` values.

Reply:
Do you know if the [`"randomize"` feature](https://github.com/rust-lang/rust/blob/9649706eada1b2c68cf6504356efb058f68ad739/compiler/rustc_abi/src/layout.rs#L981-L992) is enabled for nightly compilers? If I compile with `-Zrandomize-layout` before/after this PR both builds work, so it seems like "no" but I could also be getting unlucky with seeds

Reply:
It's part of the default features, so it should be available:

https://github.com/rust-lang/rust/blob/9649706eada1b2c68cf6504356efb058f68ad739/compiler/rustc_abi/Cargo.toml#L20

Reply:
the error is pointing at 215, so it's not the latest, but it doesn't look like the file changed since then.

```
   Compiling wast v215.0.0
error[E0080]: evaluation of constant value failed
    --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/wast-215.0.0/src/lexer.rs:65:5
     |
65   |     assert!(std::mem::size_of::<Token>() <= std::mem::size_of::<u64>() * 2);
     |     |
     |     |
     |     the evaluated program panicked at 'assertion failed: std::mem::size_of::<Token>() <= std::mem::size_of::<u64>() * 2', /cargo/registry/src/index.crates.io-6f17d22bba15001f/wast-215.0.0/src/lexer.rs:65:5
     |
    ::: /checkout/library/core/src/macros/mod.rs:1596:5
```

Reply:
tried a local checkout, it failed on the first try for me, I didn't even have to iterate through seeds.

```
$ RUSTFLAGS="-Zrandomize-layout" cargo +nightly check
    Checking wast v216.0.0 (/home/the8472/git/wasm-tools/crates/wast)
error[E0080]: evaluation of constant value failed
  --> crates/wast/src/lexer.rs:65:5
   |
65 |     assert!(std::mem::size_of::<Token>() <= std::mem::size_of::<u64>() * 2);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: std::mem::size_of::<Token>() <= std::mem::size_of::<u64>() * 2', crates/wast/src/lexer.rs:65:5
   |
   = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0080`.
error: could not compile `wast` (lib) due to 1 previous error
```

Reply:
Ah ok yeah it took a few tries but I was able to get the crate to fail the assertion before this commit as well. After this commit though I can't get it to fail yet, I'll probably just move this to a runtime test in the meantime.